### PR TITLE
Make HPACKHeaders hashable.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

HPACKHeaders is equatable, and it rarely makes sense to have a type that
is equatable but not hashable. The definition of equatability is pretty
limited here (see #342) but we can define a definition of hashability
consistent with it.

Modifications:

- Make HPACKHeaders hashable.

Result:

People can store HPACKHeaders in Sets

### CODE REVIEW

File: Sources/NIOHPACK/HPACKHeader.swift [hunk 1]
```diff
@@ -394,6 +394,9 @@ extension HPACKHeaders: CustomStringConvertible {
     }
 }
 
+// NOTE: This is a bad definition of equatable and hashable. In particular, both order and
+// indexability are ignored. We're sort of stuck with this behaviour now and cannot change it,
+// but in the next major version we should strongly consider revising it.
```
Comment (Davidde94) [inline Sources/NIOHPACK/HPACKHeader.swift:399]: Should we link a GH issue?

File: Sources/NIOHPACK/HPACKHeader.swift [hunk 2]
```diff
@@ -416,6 +419,22 @@ extension HPACKHeaders: Equatable {
...
+extension HPACKHeaders: Hashable {
+    @inlinable
+    public func hash(into hasher: inout Hasher) {
+        // Discriminator, to indicate that this is a collection. This improves the performance
+        // of Sets and Dictionaries that include collections of HPACKHeaders by reducing hash collisions.
+        hasher.combine(self.count)
+
+        // This emulates the logic used in equatability, but we sort it to ensure that
+        // we hash equivalently.
+        let names = Set(self.headers.map { $0.name }).sorted()
```
Comment (Davidde94) [inline Sources/NIOHPACK/HPACKHeader.swift:431]: Why do we need to load into a `Set` to then immediately extract?

**Review by Davidde94 — APPROVED** [review]
Minor question, otherwise looks good

File: Sources/NIOHPACK/HPACKHeader.swift (hunk 2 above)
Comment (Lukasa) [reply Sources/NIOHPACK/HPACKHeader.swift:431]: Flattening uniqueness: it's possible to have multiple entries with the same name. This is what happens above in the equatable implementation.

File: Sources/NIOHPACK/HPACKHeader.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/NIOHPACK/HPACKHeader.swift:399]: Oh, I think this comment is just unnecessarily aggressive.

File: Sources/NIOHPACK/HPACKHeader.swift [hunk 3]
```diff
@@ -416,6 +419,22 @@ extension HPACKHeaders: Equatable {
...
+        for name in names {
+            hasher.combine(self[name].sorted())
```
Comment (dnadoba) [inline Sources/NIOHPACK/HPACKHeader.swift:433]: Does the order of calls to `Hasher.combine(_:)` matter?

File: Sources/NIOHPACK/HPACKHeader.swift (hunk 2 above)
Comment (dnadoba) [reply Sources/NIOHPACK/HPACKHeader.swift:431]: We can remove one allocation if we lazily map over the headers to extract the name
```suggestion
        let names = Set(self.headers.lazy.map { $0.name }).sorted()
```

**Review by dnadoba — COMMENTED** [review]
One perf and one question, otherwise LGTM

**Pull request merged**